src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs

```diff
@@ -898,6 +878,138 @@ public static int Sum<T>(this ImmutableArray<T> items, Func<T, int> selector)
             return dictionary;
         }
 
+        internal static void AddToMultiValueDictionaryBuilder<K, T>(Dictionary<K, object> accumulator, K key, T item)
+            where K : notnull
+            where T : notnull
+        {
+            if (accumulator.TryGetValue(key, out var existingValueOrArray))
+            {
+                if (existingValueOrArray is ArrayBuilder<T> arrayBuilder)
+                {
+                    // Already a builder in the accumulator, just add to that.
+                }
+                else
+                {
+                    // Just a single value in the accumulator so far.  Convert to using a builder.
+                    arrayBuilder = ArrayBuilder<T>.GetInstance(capacity: 2);
+                    arrayBuilder.Add((T)existingValueOrArray);
+                    accumulator[key] = arrayBuilder;
+                }
+
+                arrayBuilder.Add(item);
+            }
+            else
+            {
+                // Nothing in the dictionary so far.  Add the item directly.
+                accumulator.Add(key, item);
+            }
+        }
+
+        internal static void CreateNameToMembersMap
+            <TNamespaceOrTypeSymbol, TNamedTypeSymbol, TNamespaceSymbol>
+            (Dictionary<string, object> dictionary, Dictionary<String, ImmutableArray<TNamespaceOrTypeSymbol>> result)
+            where TNamespaceOrTypeSymbol : class
+            where TNamedTypeSymbol : class, TNamespaceOrTypeSymbol
+            where TNamespaceSymbol : class, TNamespaceOrTypeSymbol
```

Comment:
```suggestion
            where TNamedTypeSymbol : class, TNamespaceOrTypeSymbol
            where TNamespaceSymbol : class, TNamespaceOrTypeSymbol
```

are the class constraints necessary here if this depend on a type parameter already with that constraint?

Reply:
> are the class constraints necessary here if this depend on a type parameter already with that constraint?

Yes, interfaces satisfy `class` constraint. An interface as a constraint doesn't mean `class`.

Reply:
my question is if restating the class constraint is necessary, or if that is picked up by the transitive type parameter constraint here.

Reply:
> my question is if restating the class constraint is necessary, or if that is picked up by the transitive type parameter constraint here.

And I answered this question, I think. With an explanation why a `class` constraint is not transitive.